pkg/querier/batch/batch.go

```diff
@@ -83,7 +83,13 @@ func newIteratorAdapter(underlying iterator) chunkenc.Iterator {
 
 // Seek implements storage.SeriesIterator.
 func (a *iteratorAdapter) Seek(t int64) bool {
-	// Optimisation: see if the seek is within the current batch.
+	// Optimisations: see if the seek is before or within the current batch.
+	if a.curr.Length > 0 && t <= a.curr.Timestamps[0] {
```

Comment:
I'm afraid this optimization is not correct. If a.Next() has already returned `false`, calling `Seek(0)` would still return `true`, indicating that there are more samples available, which is not true.

You can see this by adding more `sut.Next()` calls into new `TestSeekCorrectlyDealWithSinglePointChunks` test.

Reply:
Nice catch, let me fix that.

Reply:
I think can should address issue where "we can go backward within a batch" separately, because it is the behaviour before this change, and more time is needed to analyze if there are code that depends on the behaviour. 

I guess for now it's safe to assume caller will mostly call this method with non-decreasing timestamp?

Reply:
I think that call like `Seek(0)` can simply be used to check if Iterator still has any samples, and it should work correctly.

However, it seems to be broken without your changes as well :-( I would still prefer to not introduce new bug here, but there is more fixing to be done.

Reply:
I put in a ` a.curr.Index < a.curr.Length` check into the first if statement. I couldn't find an easy way to check if the underlying iterator still has more samples, because there are no methods like`HasNext()` for the `batch.iterator` interface, which does not change the state of the iterator. It looks to me like I might need to introduce a state variable for `batch.iteratorAdapter` to keep track of the if-iterator-reached-the-end state. But I really think we should do that in separate PR to minimize amount of logical changes in this PR as I don't 100% understand how all the iterator ties together yet.

Also, I tried to modify the `TestSeekCorrectlyDealWithSinglePointChunks` by adding the following

```go
for sut.Next() {
}

require.False(t, sut.Seek(0))

However the test would not pass. For some reason the current call to `a.underlying.Seek(t, a.batchSize)` would still return true. It seems like current chain of `Seek()` calls implements a "circular" iterator :)

I do some quick tracing and it seems like `chunk/encoding/prometheus_chunk.go`'s `FindAtOrAfter` function resets it's internal iterator.

So yea, that's why I didn't add new test; I think adding a test to verify existing incorrect behaviour would be misleading for the reader :)